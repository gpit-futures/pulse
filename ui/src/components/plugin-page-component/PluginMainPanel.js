import React, { PureComponent } from 'react';
import _ from 'lodash/fp';

import PTButton from '../ui-elements/PTButton/PTButton';
import SortableTable from '../containers/SortableTable/SortableTable';
import PaginationBlock from '../presentational/PaginationBlock/PaginationBlock';
import Spinner from '../ui-elements/Spinner/Spinner';

export default class PluginMainPanel extends PureComponent {
  static defaultProps = {
    listPerPageAmount: 10,
    emptyDataMessage: 'No list'
  };
  getClinicalNotesOnFirstPage = (list) => {
    const { listPerPageAmount, offset } = this.props;

    return (_.size(list) > listPerPageAmount
      ? _.slice(offset, offset + listPerPageAmount)(list)
      : list)
  };

  shouldHavePagination = list => _.size(list) > this.props.listPerPageAmount;

  render() {
    const { headers, resourceData, emptyDataMessage, onHeaderCellClick, onCellClick, columnNameSortBy, sortingOrder, filteredData, totalEntriesAmount, offset, setOffset, isBtnCreateVisible, onCreate, listPerPageAmount, isLoading, id } = this.props;
    const listOnFirstPage = _.flow(this.getClinicalNotesOnFirstPage)(filteredData);
    const isPagination = this.shouldHavePagination(filteredData);

    return (
      <div className="panel-body">
        <SortableTable
          headers={headers}
          data={listOnFirstPage}
          resourceData={resourceData}
          emptyDataMessage={emptyDataMessage}
          onHeaderCellClick={onHeaderCellClick}
          onCellClick={onCellClick}
          columnNameSortBy={columnNameSortBy}
          sortingOrder={sortingOrder}
          id={id}
        />
        {isLoading ? <Spinner /> : null }
        {isPagination || isBtnCreateVisible ?
          <div className="panel-control">
            <div className="wrap-control-group">
              { isPagination &&
              <div className="control-group with-indent left">
                <PaginationBlock
                  entriesPerPage={listPerPageAmount}
                  totalEntriesAmount={totalEntriesAmount}
                  offset={offset}
                  setOffset={setOffset}
                />
              </div>
              }
              {isBtnCreateVisible && <div className="control-group with-indent right">
                <PTButton className="btn btn-success btn-inverse btn-create" onClick={() => onCreate()}>
                  <i className="btn-icon fa fa-plus" />
                  <span className="btn-text"> Create</span>
                </PTButton>
              </div>}
            </div>
          </div>
          : null
        }

      </div>
    )
  }
}
