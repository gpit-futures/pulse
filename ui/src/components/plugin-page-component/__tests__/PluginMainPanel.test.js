import React from 'react';
import { configure, shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-15';

configure({ adapter: new Adapter() });

import PluginMainPanel from '../PluginMainPanel';

const testProps = {
  onHeaderCellClick: () => {},
  onCellClick: () => {},
  onCreate: () => {},
  setOffset: () => {},
  headers: [{
    key: 'clinicalNotesType',
    title: 'Type',
    width: '23%',
  }],
  resourceData: [1, 2, 3, 4, 5, 6, 7, 8, 9],
  filteredData: [1, 3, 5, 7, 9],
  columnNameSortBy: 'nameColumn',
  sortingOrder: 'asc',
  totalEntriesAmount: 5,
  offset: 10,
  id: 25,
  listPerPageAmount: 5,
  emptyDataMessage: 'My list is empty',
  emptyDataMessageDefault: 'No list',
};


describe('Component <PluginMainPanel />', () => {
  it('should renders with all props correctly', () => {
    let tree;
    const component = shallow(
      <PluginMainPanel
        onHeaderCellClick={testProps.onHeaderCellClick}
        onCellClick={testProps.onCellClick}
        onCreate={testProps.onCreate}
        headers={testProps.headers}
        resourceData={testProps.resourceData}
        columnNameSortBy={testProps.columnNameSortBy}
        sortingOrder={testProps.sortingOrder}
        filteredData={testProps.filteredData}
        totalEntriesAmount={testProps.totalEntriesAmount}
        offset={testProps.offset}
        setOffset={testProps.setOffset}
        id={testProps.id}
        isBtnCreateVisible={false}
        isLoading={false}
      />);

    expect(component.find('SortableTable')).toHaveLength(1);
    expect(component.find('Spinner')).toHaveLength(0);
    expect(component.find('.panel-control')).toHaveLength(0);
    expect(component.find('PaginationBlock')).toHaveLength(0);
    expect(component.find('PTButton')).toHaveLength(0);

    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    expect(component.instance().props.onHeaderCellClick).toEqual(testProps.onHeaderCellClick);
    expect(component.instance().props.onCellClick).toEqual(testProps.onCellClick);
    expect(component.instance().props.onCreate).toEqual(testProps.onCreate);
    expect(component.instance().props.headers).toEqual(testProps.headers);
    expect(component.instance().props.resourceData).toEqual(testProps.resourceData);
    expect(component.instance().props.columnNameSortBy).toEqual(testProps.columnNameSortBy);
    expect(component.instance().props.sortingOrder).toEqual(testProps.sortingOrder);
    expect(component.instance().props.filteredData).toEqual(testProps.filteredData);
    expect(component.instance().props.totalEntriesAmount).toEqual(testProps.totalEntriesAmount);
    expect(component.instance().props.offset).toEqual(testProps.offset);
    expect(component.instance().props.setOffset).toEqual(testProps.setOffset);
    expect(component.instance().props.id).toEqual(testProps.id);
    expect(component.instance().props.isBtnCreateVisible).toEqual(false);
    expect(component.instance().props.isLoading).toEqual(false);

    expect(component.instance().props.listPerPageAmount).toEqual(10);
    expect(component.instance().props.emptyDataMessage).toEqual(testProps.emptyDataMessageDefault);

    component.setProps({
      listPerPageAmount: testProps.listPerPageAmount,
      emptyDataMessage: testProps.emptyDataMessage,
    });

    expect(component.instance().props.listPerPageAmount).toEqual(testProps.listPerPageAmount);
    expect(component.instance().props.emptyDataMessage).toEqual(testProps.emptyDataMessage);
  });

  it('should renders with pagination component correctly', () => {
    let tree;
    const component = shallow(
      <PluginMainPanel
        onHeaderCellClick={testProps.onHeaderCellClick}
        onCellClick={testProps.onCellClick}
        onCreate={testProps.onCreate}
        headers={testProps.headers}
        resourceData={testProps.resourceData}
        columnNameSortBy={testProps.columnNameSortBy}
        sortingOrder={testProps.sortingOrder}
        filteredData={testProps.filteredData}
        totalEntriesAmount={testProps.totalEntriesAmount}
        setOffset={testProps.setOffset}
        offset={1}
        listPerPageAmount={1}
        id={testProps.id}
        isBtnCreateVisible={false}
        isLoading={false}
      />);

    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    expect(component.find('.panel-control')).toHaveLength(1);
    expect(component.find('PaginationBlock')).toHaveLength(1);
  });

  it('should renders with spinner and create button components correctly', () => {
    let tree;
    const component = shallow(
      <PluginMainPanel
        onHeaderCellClick={testProps.onHeaderCellClick}
        onCellClick={testProps.onCellClick}
        onCreate={testProps.onCreate}
        headers={testProps.headers}
        resourceData={testProps.resourceData}
        columnNameSortBy={testProps.columnNameSortBy}
        sortingOrder={testProps.sortingOrder}
        filteredData={testProps.filteredData}
        totalEntriesAmount={testProps.totalEntriesAmount}
        setOffset={testProps.setOffset}
        offset={1}
        listPerPageAmount={1}
        id={testProps.id}
        isBtnCreateVisible
        isLoading
      />);

    tree = toJson(component);
    expect(tree).toMatchSnapshot();

    expect(component.find('.panel-control')).toHaveLength(1);
    expect(component.find('Spinner')).toHaveLength(1);
    expect(component.find('PTButton')).toHaveLength(1);

    component.find('PTButton').at(0).simulate('click');
  });
});
