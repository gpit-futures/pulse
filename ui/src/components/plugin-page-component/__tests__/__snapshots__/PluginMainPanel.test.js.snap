// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <PluginMainPanel /> should renders with all props correctly 1`] = `
<div
  className="panel-body"
>
  <SortableTable
    columnNameSortBy="nameColumn"
    data={
      Array [
        1,
        3,
        5,
        7,
        9,
      ]
    }
    emptyDataMessage="No list"
    headers={
      Array [
        Object {
          "key": "clinicalNotesType",
          "title": "Type",
          "width": "23%",
        },
      ]
    }
    id={25}
    onCellClick={[Function]}
    onHeaderCellClick={[Function]}
    resourceData={
      Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
      ]
    }
    sortingOrder="asc"
    table=""
  />
</div>
`;

exports[`Component <PluginMainPanel /> should renders with pagination component correctly 1`] = `
<div
  className="panel-body"
>
  <SortableTable
    columnNameSortBy="nameColumn"
    data={
      Array [
        3,
      ]
    }
    emptyDataMessage="No list"
    headers={
      Array [
        Object {
          "key": "clinicalNotesType",
          "title": "Type",
          "width": "23%",
        },
      ]
    }
    id={25}
    onCellClick={[Function]}
    onHeaderCellClick={[Function]}
    resourceData={
      Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
      ]
    }
    sortingOrder="asc"
    table=""
  />
  <div
    className="panel-control"
  >
    <div
      className="wrap-control-group"
    >
      <div
        className="control-group with-indent left"
      >
        <PaginationBlock
          entriesPerPage={1}
          offset={1}
          setOffset={[Function]}
          totalEntriesAmount={5}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Component <PluginMainPanel /> should renders with spinner and create button components correctly 1`] = `
<div
  className="panel-body"
>
  <SortableTable
    columnNameSortBy="nameColumn"
    data={
      Array [
        3,
      ]
    }
    emptyDataMessage="No list"
    headers={
      Array [
        Object {
          "key": "clinicalNotesType",
          "title": "Type",
          "width": "23%",
        },
      ]
    }
    id={25}
    onCellClick={[Function]}
    onHeaderCellClick={[Function]}
    resourceData={
      Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
      ]
    }
    sortingOrder="asc"
    table=""
  />
  <Spinner />
  <div
    className="panel-control"
  >
    <div
      className="wrap-control-group"
    >
      <div
        className="control-group with-indent left"
      >
        <PaginationBlock
          entriesPerPage={1}
          offset={1}
          setOffset={[Function]}
          totalEntriesAmount={5}
        />
      </div>
      <div
        className="control-group with-indent right"
      >
        <PTButton
          className="btn btn-success btn-inverse btn-create"
          onClick={[Function]}
        >
          <i
            className="btn-icon fa fa-plus"
          />
          <span
            className="btn-text"
          >
             Create
          </span>
        </PTButton>
      </div>
    </div>
  </div>
</div>
`;
